packages/vscode-extension/src/project/project.ts

```diff
@@ -117,33 +117,48 @@ export class Project implements Disposable, MetroDelegate, ProjectInterface {
     this.updateProjectState({ status: "incrementalBundleError" });
   }
 
+
+  private selectInitialDevice(devices: DeviceInfo[]) {
+    const lastDeviceId = extensionContext.workspaceState.get<string | undefined>(
+      LAST_SELECTED_DEVICE_KEY
+    );
+    let device = devices.find((item) => item.id === lastDeviceId);
+    if (!device && devices.length > 0) {
+      device = devices[0];
+    }
+    if (device) {
+      this.selectDevice(device);
+      return true;
+    }
+    this.updateProjectState({
+      selectedDevice: undefined,
+    });
+    return false;
+  }
+
   /**
    * This method tried to select the last selected device from devices list.
    * If the device list is empty, we wait until we can select a device.
    */
   private async trySelectingInitialDevice() {
-    const selectInitialDevice = (devices: DeviceInfo[]) => {
-      const lastDeviceId = extensionContext.workspaceState.get<string | undefined>(
-        LAST_SELECTED_DEVICE_KEY
-      );
-      let device = devices.find((item) => item.id === lastDeviceId);
-      if (!device && devices.length > 0) {
-        device = devices[0];
-      }
-      if (device) {
-        this.selectDevice(device);
-        return true;
-      }
-      this.updateProjectState({
-        selectedDevice: undefined,
-      });
-      return false;
-    };
+  const devices = await this.deviceManager.listAllDevices();
+    if (!this.selectInitialDevice(devices)) {
+      const listener = (newDevices: DeviceInfo[]) => {
+        if (this.selectInitialDevice(newDevices)) {
+          this.deviceManager.removeListener("devicesChanged", listener);
+        }
+      };
+      this.deviceManager.addListener("devicesChanged", listener);
+    }
+  }
 
-    const devices = await this.deviceManager.listAllDevices();
-    if (!selectInitialDevice(devices)) {
+   // Try to select device from devices list different than removedDevice in single attempt.
+  private async trySelectingAnotherDevice(removedDevice: DeviceInfo) {
```

Comment:
If I'm not mistaken, this is a copy of `trySelectingInitialDevice`. Why we're getting remove device ID from `listAllDevices()` in the first place?

Reply:
I thought that we should select different device after removing, but I guess it's not necessary and we ignore those errors in debug console when removed device is selected.

Reply:
We _should_ select different device. My point is that: at the time you call `listAllDevices()` the device that was deleted shouldn't be there. Then you don't need to add unnecessary workarounds for "device that was deleted but state wasn't updated so we need to avoid selecting it again".

Reply:
I've noticed that after removing device debug console logs `"Couldn't start device session Error: Command failed with exit code 204: xcrun simctl --set /Users/USER/Library/Caches/com.swmansion.react-native-ide/Devices/iOS install ED79C082-FDEF-46E3-94A0-088E0AE337AD /Users/USER/Library/Developer/Xcode/DerivedData/exporouterexample-fdhmhyxcaikoqicwmzpzgpagpcxx/Build/Products/Debug-iphonesimulator/exporouterexample.app..."` so my guess was that it selected the removed device. What is the reason that causes that log?